ecc/bls12-377/fr/fft/bitreverse.go

```diff
@@ -0,0 +1,574 @@
+	logTileSize := deriveLogTileSize(logN)
+	logBLen := logN - 2*logTileSize
+	bLen := uint64(1) << logBLen
+	bShift := logBLen + logTileSize
+	tileSize := uint64(1) << logTileSize
+
+	// rough idea;
+	// bit reversal permutation naive implementation may have some cache associativity issues,
+	// since we are accessing elements by strides of powers of 2.
+	// on large inputs, this is noticeable and can be improved by using a t buffer.
+	// idea is for t buffer to be small enough to fit in cache.
+	// in the first inner loop, we copy the elements of v into t in a bit-reversed order.
+	// in the subsequent inner loops, accesses have much better cache locality than the naive implementation.
+	// hence even if we apparently do more work (swaps / copies), we are faster.
+	//
+	// on arm64 (and particularly on M1 macs), this is not noticeable, and the naive implementation is faster,
+	// in most cases.
+	// on x86 (and particularly on aws hpc6a) this is noticeable, and the t buffer implementation is faster (up to 3x).
+	//
+	// optimal choice for the tile size is cache dependent; in theory, we want the t buffer to fit in the L1 cache;
+	// in practice, a common size for L1 is 64kb, a field element is 32bytes or more.
+	// hence we can fit 2k elements in the L1 cache, which corresponds to a tile size of 2**5 with some margin for cache conflicts.
+	//
+	// for most sizes of interest, this tile size choice doesn't yield good results;
+	// we find that a tile size of 2**9 gives best results for input sizes from 2**21 up to 2**27+.
+	t := make([]fr.Element, tileSize*tileSize)
+
+	// see https://csaws.cs.technion.ac.il/~itai/Courses/Cache/bit.pdf
+	// for a detailed explanation of the algorithm.
+	for b := uint64(0); b < bLen; b++ {
+
+		for a := uint64(0); a < tileSize; a++ {
+			aRev := (bits.Reverse64(a) >> (64 - logTileSize)) << logTileSize
+			for c := uint64(0); c < tileSize; c++ {
+				idx := (a << bShift) | (b << logTileSize) | c
+				t[aRev|c] = v[idx]
+			}
+		}
+
+		bRev := (bits.Reverse64(b) >> (64 - logBLen)) << logTileSize
+
+		for c := uint64(0); c < tileSize; c++ {
+			cRev := ((bits.Reverse64(c) >> (64 - logTileSize)) << bShift) | bRev
+			for aRev := uint64(0); aRev < tileSize; aRev++ {
+				a := bits.Reverse64(aRev) >> (64 - logTileSize)
+				idx := (a << bShift) | (b << logTileSize) | c
+				idxRev := cRev | aRev
+				if idx < idxRev {
+					tIdx := (aRev << logTileSize) | c
+					v[idxRev], t[tIdx] = t[tIdx], v[idxRev]
+				}
+			}
+		}
+
+		for a := uint64(0); a < tileSize; a++ {
+			aRev := bits.Reverse64(a) >> (64 - logTileSize)
+			for c := uint64(0); c < tileSize; c++ {
+				cRev := (bits.Reverse64(c) >> (64 - logTileSize)) << bShift
+				idx := (a << bShift) | (b << logTileSize) | c
+				idxRev := cRev | bRev | aRev
+				if idx < idxRev {
+					tIdx := (aRev << logTileSize) | c
+					v[idx], t[tIdx] = t[tIdx], v[idx]
+				}
+			}
+		}
+	}
+}
+
+func bitReverseCobra(v []fr.Element) {
+	switch len(v) {
```

Comment:
are these empirical results?

Reply:
in a sense; these methods are just generated with constant sized arrays and offsets for specific sizes.
Below a certain threshold, the naive version performs well, and after that threshold (2**27) I just didn't bother generating the methods since I don't think it's very common to bit reverse 270M+ vectors (but maybe...)